src/core/net/srp_server.cpp

```diff
@@ -1667,25 +1657,13 @@ void Server::Service::Description::TakeResourcesFrom(Description &aDescription)
 
 Error Server::Service::Description::SetTxtDataFromMessage(const Message &aMessage, uint16_t aOffset, uint16_t aLength)
 {
-    Error    error = kErrorNone;
-    uint8_t *txtData;
-
-    txtData = static_cast<uint8_t *>(Heap::CAlloc(1, aLength));
-    VerifyOrExit(txtData != nullptr, error = kErrorNoBufs);
-
-    VerifyOrExit(aMessage.ReadBytes(aOffset, txtData, aLength) == aLength, error = kErrorParse);
-    VerifyOrExit(Dns::TxtRecord::VerifyTxtData(txtData, aLength, /* aAllowEmpty */ false), error = kErrorParse);
+    Error error;
 
-    Heap::Free(mTxtData);
-    mTxtData   = txtData;
-    mTxtLength = aLength;
+    SuccessOrExit(error = mTxtData.SetFrom(aMessage, aOffset, aLength));
+    VerifyOrExit(Dns::TxtRecord::VerifyTxtData(mTxtData.GetBytes(), mTxtData.GetLength(), /* aAllowEmpty */ false),
+                 error = kErrorParse);
```

Comment:
Is there any reason to keep the heap data around when `VerifyTxtData()` fails?

Reply:
Added a check at the end to free the `mTxtData` (if there is `error`) but in general I think it may be ok without it as well.

The `error` here should propagate up in the call chain, causing the `Host` entry which contains this `Service` to be freed which in turn frees the service and its description and then the `mTxtData`.